src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMStep.java

```diff
@@ -19,129 +19,82 @@
 import hudson.scm.SCM;
 import hudson.util.FormValidation;
 import hudson.util.ListBoxModel;
+import org.apache.commons.lang3.StringUtils;
 import org.jenkinsci.Symbol;
 import org.jenkinsci.plugins.workflow.steps.scm.SCMStep;
-import org.kohsuke.stapler.AncestorInPath;
-import org.kohsuke.stapler.DataBoundConstructor;
-import org.kohsuke.stapler.HttpResponse;
-import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.*;
 import org.kohsuke.stapler.verb.POST;
 
 import javax.annotation.Nullable;
 import javax.inject.Inject;
-import java.util.List;
-import java.util.Optional;
-import java.util.UUID;
-import java.util.logging.Logger;
+import java.util.*;
 
 import static java.lang.String.format;
 import static java.util.Collections.emptyList;
 import static java.util.Collections.singletonList;
+import static java.util.Objects.requireNonNull;
 import static org.apache.commons.lang3.StringUtils.isBlank;
 
 public class BitbucketSCMStep extends SCMStep {
 
-    private static final Logger LOGGER = Logger.getLogger(BitbucketSCMStep.class.getName());
-
-    private final List<BranchSpec> branches;
-    private final String cloneUrl;
-    private final String credentialsId;
-    private final String id;
-    private final String projectKey;
+    private List<BranchSpec> branches;
+    private String cloneUrl;
+    private String credentialsId;
+    private String sshCredentialsId;
+    private String id;
+    private String projectKey;
     private final String projectName;
     private final String repositoryName;
-    private final String repositorySlug;
-    private final int repositoryId;
-    private final String selfLink;
+    private String repositorySlug;
+    private int repositoryId;
+    private String selfLink;
     private final String serverId;
-    private final String mirrorName;
+    private String mirrorName;
 
     @DataBoundConstructor
```

Comment:
The main changes in the PR are here. The SCM step constructor has been reduced to just the mandatory fields, with the optional fields (credentials, mirror name etc.) now being provided in data bound setters. This permits those fields to be absent in the groovy syntax- after the constructor is complete Stapler finds matching setters for each one, and discards the rest. Note however this means most initialization occurs in the `createSCM` method, which I will argue as a good thing, because this constructor is only initialized in two different situations:
- When the snippet is run, in which case createSCM is invoked immediately after the constructor
- When the snippet is generated. We don't call createSCM because we don't need to- the snippet only needs those fields to have values